File: android/src/main/java/com/reactlibrary/NoCodesModule.java
package com.reactlibrary;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.module.annotations.ReactModule;
import io.qonversion.sandwich.NoCodesSandwich;
import io.qonversion.sandwich.NoCodesEventListener;

import java.util.HashMap;
import java.util.Map;

public class NoCodesModule extends ReactContextBaseJavaModule {
    private final ReactApplicationContext reactContext;
    private final NoCodesSandwich noCodesSandwich;
    private final NoCodesEventListener noCodesEventListener;
    private DeviceEventManagerModule.RCTDeviceEventEmitter eventEmitter = null;

    public NoCodesModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.noCodesSandwich = new NoCodesSandwich();

        this.noCodesEventListener = (event, payload) -> {
            WritableMap payloadMap = null;
            if (payload != null) {
                payloadMap = EntitiesConverter.convertMapToWritableMap(payload);
            }

            sendEvent(event.getKey(), payloadMap);
        };
    }

    @Override
    public void initialize() {
        super.initialize();

        eventEmitter = getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class);
    }

    @NonNull
    @Override
    public String getName() {
        return "RNNoCodes";
    }

    @ReactMethod
    public void initialize(String projectKey) {
        noCodesSandwich.initialize(reactContext, projectKey, null, null, null);
        noCodesSandwich.setDelegate(noCodesEventListener);
        noCodesSandwich.setScreenCustomizationDelegate();
    }

    @ReactMethod
    public void setScreenPresentationConfig(ReadableMap configData, String contextKey, final Promise promise) {
        try {
            Map<String, Object> config = EntitiesConverter.convertReadableMapToHashMap(configData);
            noCodesSandwich.setScreenPresentationConfig(config, contextKey);
            Utils.resolveWithSuccess(promise);
        } catch (Exception e) {
            promise.reject(e);
        }
    }

    @ReactMethod
    public void showScreen(String contextKey, final Promise promise) {
        try {
            noCodesSandwich.showScreen(contextKey);
            Utils.resolveWithSuccess(promise);
        } catch (Exception e) {
            promise.reject(e);
        }
    }

    @ReactMethod
    public void close(final Promise promise) {
        try {
            noCodesSandwich.close();
            Utils.resolveWithSuccess(promise);
        } catch (Exception e) {
            promise.reject(e);
        }
    }

    private void sendEvent(String eventName, WritableMap params) {
        if (eventEmitter != null) {
            eventEmitter.emit(eventName, params);
        }
    }
}

File: android/src/main/java/com/qonversion/reactnativesdk/NoCodesModule.kt
package com.qonversion.reactnativesdk

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeMap
import io.qonversion.sandwich.BridgeData
import io.qonversion.sandwich.NoCodesSandwich
import io.qonversion.sandwich.NoCodesEventListener
import io.qonversion.sandwich.NoCodesPurchaseDelegateBridge

@ReactModule(name = NoCodesModule.NAME)
class NoCodesModule(private val reactContext: ReactApplicationContext) : NativeNoCodesModuleSpec(reactContext) {
    private val noCodesSandwich: NoCodesSandwich = NoCodesSandwich()
    private val noCodesEventListener: NoCodesEventListener = object : NoCodesEventListener {
        override fun onNoCodesEvent(event: NoCodesEventListener.Event, payload: BridgeData?) {
            val eventMap: WritableMap = WritableNativeMap()
            eventMap.putString("name", event.key)
            payload?.let {
                val payloadMap = EntitiesConverter.convertMapToWritableMap(it)

                eventMap.putMap("payload", payloadMap)
            }
            emitOnNoCodeEvent(eventMap)
        }
    }
    private val purchaseDelegate: NoCodesPurchaseDelegateBridge = object : NoCodesPurchaseDelegateBridge {
        override fun purchase(product: BridgeData) {
            val productMap = EntitiesConverter.convertMapToWritableMap(product)
            emitOnNoCodePurchase(productMap)
        }

        override fun restore() {
            emitOnNoCodeRestore()
        }
    }

    override fun getName(): String {
        return NAME
    }

    @ReactMethod
    override fun initialize(projectKey: String, source: String, version: String, proxyUrl: String?) {
        noCodesSandwich.storeSdkInfo(reactContext, source, version)
        noCodesSandwich.initialize(reactContext, projectKey, null, null, proxyUrl)
        noCodesSandwich.setDelegate(noCodesEventListener)
        noCodesSandwich.setScreenCustomizationDelegate()
    }

    @ReactMethod
    override fun setScreenPresentationConfig(configData: ReadableMap, contextKey: String?, promise: Promise) {
        try {
            val config: Map<String, Any> = EntitiesConverter.convertReadableMapToHashMap(configData)
            noCodesSandwich.setScreenPresentationConfig(config, contextKey)
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject(e)
        }
    }

    @ReactMethod
    override fun showScreen(contextKey: String, promise: Promise) {
        try {
            noCodesSandwich.showScreen(contextKey)
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject(e)
        }
    }

    @ReactMethod
    override fun close(promise: Promise) {
        try {
            noCodesSandwich.close()
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject(e)
        }
    }

    @ReactMethod
    override fun setPurchaseDelegate() {
        noCodesSandwich.setPurchaseDelegate(purchaseDelegate)
    }

    @ReactMethod
    override fun delegatedPurchaseCompleted() {
        noCodesSandwich.delegatedPurchaseCompleted()
    }

    @ReactMethod
    override fun delegatedPurchaseFailed(errorMessage: String) {
        noCodesSandwich.delegatedPurchaseFailed(errorMessage)
    }

    @ReactMethod
    override fun delegatedRestoreCompleted() {
        noCodesSandwich.delegatedRestoreCompleted()
    }

    @ReactMethod
    override fun delegatedRestoreFailed(errorMessage: String) {
        noCodesSandwich.delegatedRestoreFailed(errorMessage)
    }

    companion object {
        const val NAME = "RNNoCodes"
    }
}
